File: BDS/TypeScript/Customer.ts
class DateDiff {

    static inDays(d1: Date, d2: Date): number {
        var t2 = d2.getTime();
        var t1 = d1.getTime();

        return (t2 - t1) / (24 * 3600 * 1000);
    }

    static inWeeks(d1: Date, d2: Date): number {
        let t2 = d2.getTime();
        let t1 = d1.getTime();

        return (t2 - t1) / (24 * 3600 * 1000 * 7);
    }

    static inMonths(d1: Date, d2: Date): number {
        let d1Y = d1.getFullYear();
        let d2Y = d2.getFullYear();
        let d1M = d1.getMonth();
        let d2M = d2.getMonth();

        return (d2M + 12 * d2Y) - (d1M + 12 * d1Y);
    }

    static inYears(d1: Date, d2: Date): number {
        return d2.getFullYear() - d1.getFullYear();
    }
}

export class Customer {
    constructor(public customerName: string, public customerSince: Date, public dateOfLastPurchase: Date, public birthday: Date) {
        this.customerName = customerName;
        this.customerSince = customerSince;
        this.dateOfLastPurchase = dateOfLastPurchase;
        this.birthday = birthday;
    }

    toString() {
        return
        "{" +
            "Name: " + this.customerName + ", " +
            "Since: " + this.customerSince + ", " +
            "Last Purchase: " + this.dateOfLastPurchase + ", " +
            "Birthday " + this.birthday +
            "}";
    }
}

export class CustomerDiscount {
    static calculateDiscount(customer: Customer) {
        // If Long term customer (more than 10 years)
        if (DateDiff.inYears(new Date(), customer.customerSince) > 10)
            return 0.20;

        // Happy Birthday
        if (new Date().getDate() == customer.birthday.getDate() && new Date().getMonth() == customer.birthday.getMonth())
            return .15;

        // If last purchase in 30 days
        if (DateDiff.inMonths(new Date(), customer.dateOfLastPurchase) < 30)
            return .08;
        return 0;
    }
}

File: BDS/Java/Swizilla/src/Customer.java
import java.time.LocalDate;
import java.time.Period;
import java.util.Date;

class Customer {

	private String customerName;
	public LocalDate customerSince;
	public LocalDate dateOfLastPurchase;
	public LocalDate birthday;

	public String getCustomerName() {
		return customerName;
	}

	private void setCustomerName(String customerName) {
		this.customerName = customerName;
	}

	public LocalDate getCustomerSince() {
		return customerSince;
	}

	private void setCustomerSince(LocalDate customerSince) {
		this.customerSince = customerSince;
	}

	public LocalDate getDateOfLastPurchase() {
		return dateOfLastPurchase;
	}

	private void setDateOfLastPurchase(LocalDate dateOfLastPurchase) {
		this.dateOfLastPurchase = dateOfLastPurchase;
	}

	public LocalDate getBirthday() {
		return birthday;
	}

	private void setBirthday(LocalDate birthday) {
		this.birthday = birthday;
	}


	public Customer(String customerName, LocalDate customerSince, LocalDate dateOfLastPurchase, LocalDate birthday) {
		this.customerName = customerName;
		this.customerSince = customerSince;
		this.dateOfLastPurchase = dateOfLastPurchase;
		this.birthday = birthday;
	}

	@Override
	public String toString()
        {
            return 
                "{" +
                "Name: " + customerName + ", " +
                "Since: " + customerSince + ", " +
                "Last Purchase: " + dateOfLastPurchase+ ", " +
                "Birthday " + birthday +
                "}";
        }
}

class CustomerDiscount {
	public static double calculateDiscount(Customer customer) {
		// If Longterm customer (more than 10 years)
		Period years = Period.between(LocalDate.now(), customer.customerSince);
		if (years.getYears() > 10)
			return 0.20;

		// Happy Birthday
		if ((LocalDate.now().getDayOfMonth() == customer.birthday.getDayOfMonth()) && (LocalDate.now().getMonth() == customer.birthday.getMonth()))
			return .15;

		// If last purchase in 30 days
		Period days = Period.between(LocalDate.now(), customer.dateOfLastPurchase);
		if (days.getDays() < 30)
			return .08;
		return 0;
	}
}